rtic-sync/src/channel.rs

```diff
@@ -77,8 +77,61 @@ impl<T, const N: usize> Channel<T, N> {
         }
     }
 
+    /// Clear any remaining items from this `Channel`.
+    pub fn clear(&mut self) {
+        for _ in self.queued_items() {}
+    }
+
+    /// Return an iterator over the still-queued items, removing them
+    /// from this channel.
+    pub fn queued_items(&mut self) -> impl Iterator<Item = T> + '_ {
+        struct Iter<'a, T, const N: usize> {
+            inner: &'a mut Channel<T, N>,
+        }
+
+        impl<T, const N: usize> Iterator for Iter<'_, T, N> {
+            type Item = T;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                let slot = self.inner.readyq.get_mut().pop_back()?;
+
+                let value = unsafe {
+                    // SAFETY: `ready` is a valid slot.
+                    let ptr = self.inner.slots.get_unchecked(slot as usize).get() as *const T;
+                    // SAFETY: `ptr` points to an initialized `T`.
+                    core::ptr::read(ptr)
+                };
+
+                // NOTE: do not `return_free_slot`, as we have mutable
+                // access to this `Channel` and no `Receiver` or `Sender`
+                // exist.
+                assert!(!self.inner.freeq.get_mut().is_full());
+                unsafe {
+                    // SAFETY: `freeq` is not ful.
+                    self.inner.freeq.get_mut().push_back_unchecked(slot);
+                }
+
+                Some(value)
+            }
+        }
+
+        Iter { inner: self }
+    }
+
     /// Split the queue into a `Sender`/`Receiver` pair.
+    ///
+    /// # Panics
+    /// This function panics if there are items in this channel while splitting.
+    ///
+    /// Call [`Channel::clear`] to clear all items from it, or [`Channel::queued_items`] to retrieve
+    /// an iterator that yields the values.
     pub fn split(&mut self) -> (Sender<'_, T, N>, Receiver<'_, T, N>) {
+        assert!(
+            self.readyq.get_mut().is_empty(),
```

Comment:
Since we have `&mut channel` here, can't we do `self.clear()`?

Reply:
Yes. When writing this felt weird to drop the items in the queue on `split`, but and not any earlier. Since most callers won't ever re-`split()` their queue I figured we could get away with never calling `clear` if they do not.

If you think we should `clear()`, should we do it in `Receiver::drop` too? It just feels kind of arbitrary to `drop` leftover items on `split()` and not earlier (i.e. `Receiver::drop`), but I have no clue what the usual thing to do is, here.

Reply:
Yeah, true. Sounds like `Receiver::drop` is where it should happen.